File: rules/httponly.js
const fs = require('fs');
const url = require('url');
const vscode = require('vscode');

var smell = {

    detect : (fileName, token) => {
        try{
            const MSG = 'possible use of HTTP without TLS'
        
            const WARNING_MSG = MSG+' at line '+ lineno;
            const WARNING_MSG_ON_RETURN = token.hasOwnProperty("returnLine") ? WARNING_MSG+ token.returnLine : null;
    
            if(token.hasOwnProperty("line")) var lineno = token.line;
            if(token.hasOwnProperty("args")) var args = token.args;
    
            const httpLibs = [  'httplib.urlretrieve','urllib.request.urlopen','urllib.urlopen','urllib2.urlopen','requests.get', 
                                'requests.post','urllib.request.Request','httplib.HTTPConnection','httplib2.Http.request'
                            ]
    
            const newHttpLibs = ['urllib3.PoolManager.request']
            
            if(token.type == 'variable' && token.value != null && token.valueSrc == 'initialization'){
                if(smell.isValidHTTPURL(token.value)) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
            }
            if(token.type == 'variable' && token.hasOwnProperty('funcKeywords')){
                
                token.funcKeywords.forEach(keyword => {
                    if(keyword[1] != null && smell.isValidHTTPURL(keyword[1])){
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
                })
            }
            if(token.type == 'variable' && token.hasOwnProperty('valueSrc') && token.hasOwnProperty('args')){
                if(httpLibs.includes(token.valueSrc) && token.args.length > 0 && smell.isValidHTTPURL(token.args[0])){
                    smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
    
                else if(newHttpLibs.includes(token.valueSrc) && token.args.length > 1 && smell.isValidHTTPURL(token.args[1])){
                    smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
            }
            else if(token.type == 'function_call' && token.name != null){
                if(httpLibs.includes(token.name) && token.args.length > 0 && smell.isValidHTTPURL(args[0])){
                    smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
    
                else if(newHttpLibs.includes(token.name) && token.args.length > 1 && smell.isValidHTTPURL(args[1])){
                    smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                }
            }
            else if(token.type == 'function_def' && token.hasOwnProperty('return') && token.return != null && token.hasOwnProperty('returnArgs')){
                
                for(const funcReturn of token.return){
                    if(httpLibs.includes(funcReturn) && token.returnArgs.length > 0 && smell.isValidHTTPURL(token.returnArgs[0])){
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG_ON_RETURN);
                    }
        
                    else if(newHttpLibs.includes(funcReturn) && token.returnArgs.length > 1 && smell.isValidHTTPURL(token.returnArgs[1])){
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG_ON_RETURN);
                    }
                }
    
            }
        } catch(error){
            console.log(error);
        }
    },
    isValidHTTPURL: hostURL => {
        if(typeof(hostURL) == 'string'){

            const urlPattern = 'http[s]?:\/\/(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\), ]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
            let urlRegex = new RegExp(urlPattern)

            // console.log(urlRegex.test(hostURL))

            if(urlRegex.test(hostURL) == false) return false
            
            const hostURLObj = url.parse(hostURL)
            // console.log(hostURLObj.protocol);
            if(hostURLObj.protocol == 'http:') return true
            else return false

        }
        else return false
    }      ,
    
    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]
        
        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        
        fs.appendFileSync(__dirname+'/../warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
        // fs.appendFile(__dirname+'/../logs/project_warnings.csv', fileName+","+WARNING_MSG+"\n", (err) => err ? console.log(err): "");
    }     
}

module.exports = smell;

File: py-scripts/filters/httponly.py
import re
from urllib.parse import urlparse

class HttpWithoutTLS:
    '''This is the class for detecting use of HTTP without TLS in code'''

    def __init__(self):
        self.http_libs = ['httplib.urlretrieve','urllib.request.urlopen','urllib.urlopen','urllib2.urlopen','requests.get', 
                        'requests.post','urllib.request.Request','httplib.HTTPConnection','httplib2.Http.request'
                    ]

        self.new_http_libs = ['urllib3.PoolManager.request']
        self.warning_message = 'use of HTTP without TLS'


    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"]
            if token.__contains__("type"): tokenType = token["type"]
            if token.__contains__("name"): name = token["name"]
            if token.__contains__("args"): args = token["args"]

            if tokenType == "variable" and token.__contains__('value') and token.__contains__('valueSrc'):
                if token['value'] is not None and  self.is_valid_http_url(token['value']) and token['valueSrc'] == 'initialization':
                    self.trigger_alarm(src_file, lineno)

            if tokenType == "variable" and token.__contains__('funcKeywords'):
                for keyword in token['funcKeywords']:
                    if keyword[1] is not None and self.is_valid_http_url(keyword[1]):
                        self.trigger_alarm(src_file, lineno)
                        break

            if tokenType == "variable" and token.__contains__("valueSrc") and token.__contains__("args"):
                args = token['args']
                valueSrc = token['valueSrc']

                if valueSrc in self.http_libs and len(args) > 0 and args[0] is not None and self.is_valid_http_url(args[0]):
                    self.trigger_alarm(src_file, lineno)
                
                elif valueSrc in self.new_http_libs and len(args) > 1 and args[1] is not None and self.is_valid_http_url(args[1]):
                    self.trigger_alarm(src_file, lineno)
        
            elif tokenType == "function_call" and name is not None: 
                if name in self.http_libs and len(args) > 0 and args[0] is not None and self.is_valid_http_url(args[0]):                                     
                    self.trigger_alarm(src_file, lineno)

                elif name in self.new_http_libs and len(args) > 1 and args[1] is not None and self.is_valid_http_url(args[1]):
                    self.trigger_alarm(src_file, lineno)

            elif tokenType == "function_def" and token.__contains__('return') and token.__contains__('returnArgs') and token["return"] is not None:
                returnArgs = token['returnArgs']

                for func_return in token["return"]:
                    if func_return in self.http_libs and len(returnArgs) > 0 and returnArgs[0] is not None and self.is_valid_http_url(returnArgs[0]):           
                        self.trigger_alarm(src_file, lineno)

                    elif func_return in self.new_http_libs and len(returnArgs) > 1 and returnArgs[1] is not None and self.is_valid_http_url(returnArgs[1]):           
                        self.trigger_alarm(src_file, lineno)

        except Exception as error: 
            print(str(error))


    def is_valid_http_url(self, url): 

        reg_url = re.findall(r'http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\), ]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', str(url))
        url = reg_url[0] if len(reg_url) > 0 else None
        if url is None: return False

        parsed_url = urlparse(url)

        if parsed_url.scheme == 'http': return True
        elif parsed_url.scheme == 'https': return False
        # elif parsed_url.scheme == '': return True
        else: return True
    

    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    